### PR TITLE
[WIP] Build using LLVM 3.8

### PR DESCRIPTION
**Pull request by matiasgarciaisaia**
### TODO list

- [x] Test CentOS 7 x64 ❌ _(has GLIBCXX_3.4.19, LLVM needs newer version)_
- [x] Get a CentOS i386 box - _Testing [jasonc/centos7-32bit](https://atlas.hashicorp.com/jasonc/boxes/centos7-32bit)_
- [x] Test CentOS 7 i386 ❌ _(has GLIBCXX_3.4.19, LLVM needs newer version)_
- [ ] Update Vagrantfile
- [ ] ???
- [ ] Profit!

### Rationale

As crystal-lang/crystal#4104 says, LLVM < 3.8 generates incorrect binary in release mode. It's bugged, so we should stop using it.

This PR upgrades the omnibus setup so the compiler is built using LLVM 3.8.1. The major drawback is that **it drops support for older Linux releases**. LLVM 3.8 seems to depend on libstdc++ newer than the ones that came with Debian 7 or Centos 6 (the boxes we were using until this PR), so Crystal would start to be compatible with Debian 8 or newer, and probably Centos 7 and on - have to check yet.

We are (temporarily) losing (incorrect, buggy) support for old software, in exchange for fixing bugs for newer platforms.

### In the long term

I've said it's a _temporary_ loss. In the long run, I've talked with @waj about building a purely-static compiler for each arch (osx, linux32, linux64, windows whenever we have it...) that doesn't depend on _any_ runtime at all (libstdc++, libc, _nothing_), and that compiler would be the one used by each distro to build it's own _distro dependent_ packages. In that scenario, a newer Debian would depend on the system's libstdc++/llvm/whatever, while an older Debian would link them statically anyway.

Of course, that means _a lot of work_, so that's why this [temporary solution](http://stackoverflow.com/a/778275/641451) could make sense in the mean time.

### Why care writing all of this instead of just merging and shipping?

Because **we're dropping support** for some things, and we don't know how big of an impact that means. It may be that no one cares about Debian 7, or it may be that the 90% of Crystal's user base is running Debian 7 - so it doesn't make sense to have this temporary solution.

**We need to know if this affects you**. Please try to test the PR, or at least comment if dropping support may be an issue for you. At least give this a 👍 or 👎 as a reaction.

### Things that have been done for this PR

I've upgraded Omnibus to the latest release, so it is compatible with the current stable Ruby (RVM gets to decide which Ruby to install during the Vagrant setup).

32 bit releases get the `--no-debug` automatically added so we don't hit the memory limit during the compilation.

The LLVM 3.8 builds have been made by @waj - I'm trying to reproduce (and document) how to get them generated, but that'll be a different PR whenever I get there.

### CODE REVIEW

**Comment by sdogruyol** [comment]
Go for it 👍

**Comment by chuckremes** [comment]
With Crystal at release 0.22.0 I think it's a waste of your resources to try and support older OS releases. Drop support and don't look back.

Target support for the latest releases of everything (Ubuntu 17.x, Mint 18.x, FreeBSD 11.x, OpenBSD 6.x, OS X 10.12.x, Windows 10, etc). As newer OS's come out and you are still in BETA, drop the older ones and target the latest!

When you hit 1.0 release-candidate status and people start relying on your releases, then lock this down. 

IMHO...

**Comment by drosehn** [comment]
FWIW, I know a fair number of people (including myself) who are on FreeBSD 10.x and macOS 10.10 (Yosemite), not because they want to be but because of problems getting to those newer releases.  I don't follow debian, so I can't say anything about that.  We (RPI) have a lot of systems running RHEL 6.x and RHEL 7.x, but I don't know if those releases would have any trouble with LLVM 3.8.

If you need to drop support for any of those releases to get to LLVM 3.8.1, then I'm fine with that.  But I hope crystal will attempt to support older OS releases until some significant problem ( *such as bugs in LLVM!* ) forces the question to come up.

**Comment by ysbaddaden** [comment]
This only affects the officially distributed packages. The Crystal compiler itself is still compatible with older releases; they have to upgrade `libstdc++` thought (or statically link it).

**Comment by drosehn** [comment]
After getting a little sleep, it occurs to me that crystal on my Yosemite Macs are already using LLVM 4.0.0 (via homebrew).  Or at least they claim to be in the output of `--version`.  😄

**Comment by matiasgarciaisaia** [comment]
So - I couldn't get LLVM 3.8 builds on Centos 7, due to libstdc++ incompatibilities.

I'm leaving this here for a while, focusing on upgrading the build chain anyway, and then I'll come back to see how to solve this issue. Dropping the latest Centos doesn't sound _that great_, but maybe it's just that I don't get how to avoid it.

I've just updated the PR with the latest things I've tried, should anyone else want to check and give a hand.

File: config/software/crystal.rb [hunk 1]
```diff
@@ -42,7 +42,7 @@
   command "mkdir .build", env: env
   command "echo #{Dir.pwd}", env: env
   command "cp #{Dir.pwd}/crystal-#{ohai['os']}-#{ohai['kernel']['machine']} .build/crystal", env: env
-  command "bin/crystal build src/compiler/crystal.cr --release -o #{output_bin} -D without_openssl -D without_zlib", env: env
+  command "bin/crystal build src/compiler/crystal.cr --release#{" --no-debug" if _32_bit? } -o #{output_bin} -D without_openssl -D without_zlib", env: env
```
Comment (ysbaddaden) [inline config/software/crystal.rb:45]: Why pass `--no-debug` on 32 bits targets? Is the DWARF parser failing, or?

**Review by ysbaddaden — COMMENTED** [review]
What about enabling the ARM and AArch64 backends, while we're upgrading LLVM?

**Comment by ysbaddaden** [comment]
Thought I believe LLVM 3.9 fixed some ARM issues in release mode... So maybe not enable these just yet.

**Comment by bararchy** [comment]
Not affected , also it seems most Crystal users are ether on OSx or Arch Linux

**Comment by ysbaddaden** [comment]
I ran into the libstdc++ issues with my raspbian packages. I installed LLVM 3.9 from unstable which required newer versions of libtinfo and libstdc++. I simply linked these statically, just like LLVM. It only affects the compiler anyway, an it just works, without affecting the system.

About the omnibus, I suppose the actual issue is how to get the newer versions installed, or at least have working static libraries available?

File: config/software/crystal.rb (hunk 1 above)
Comment (matiasgarciaisaia) [reply config/software/crystal.rb:45]: If we don't use `--no-debug`, we hit the 4GB memory limit for 32 bits systems - we should cross-compile, then.

**Comment by matiasgarciaisaia** [comment]
@ysbaddaden have a quick link to share about how to statically link LLVM3.8/libstdc++? That way we may not lose support.

**Comment by ysbaddaden** [comment]
LLVM is usually statically linked, though this depends on the system, otherwise `--link-static` can be passed to `llvm-config`.

For `libstdc++` and `libtinfo`, I cross compile the binary, and manually link in a qemu VM. I can pass `-Wl,-Bstatic -lstdc++ -ltinfo -Wl,-Bdynamic <any other libs>` so the linker will choose static over shared libs.

I'm not sure this can be easily integrated into the omnibus.

**Comment by matiasgarciaisaia** [comment]
The changes in this PR eventually got directly [pushed to `current` while releasing 0.23.0](https://github.com/crystal-lang/omnibus-crystal/compare/c55a21dc759f0575a5dc8eed6ee0e92d60e17f1b...d1e46859289d5f99ceda98adac20917cc27bf2c5), so I'm closing the PR.

Thanks to everybody for your inputs & support!

**Comment by solisoft** [comment]
I use centos7 and crystal 0.22 for a production script ...

**Comment by sdogruyol** [comment]
@solisoft sad to hear that 😢

**Comment by Fryguy** [comment]
I'm also concerned about the loss of CentOS 7, as we planned to use that for production release as well.

**Comment by mverzilli** [comment]
@solisoft and @Fryguy this should only affect the prebuilt packages, you should still be able to build from source. 

We'll eventually fix it, but LLVM3.5 was causing too many issues already and we needed to deprecate it otherwise we were kind of deadlocking development. 

Two things:

1. Can you make do building from source until we figure out a way to generate official CentOS 7 packages again?
2. If you can't, let's work together on a solution. If you feel comfortable sharing how you provision and setup your scripts here, please do so so we can assist. If not, drop me an email to mverzilli@manas.tech

Thank you!

**Comment by akzhan** [comment]
btw, static linkage is great benefit of Golang so I want static linking to be this default.

**Comment by RX14** [comment]
Static linking the compiler into a single binary should replace the current omnibus setup (which seems to have about 100k lines too much code to produce a tarball). The only other library we should distribute with the omnibus is `libgc`, every other library should come from the distribution when the compiler is linking. Libgc has no dependencies other than libc so it should cause very few if any issues. Issues could arise if we're linking to a library which also uses libgc but that's got to be pretty rare.

I don't think static linking compiler output should be the default but it should be an option, and distributing the crystal compiler is the perfect usecase for it.

**Comment by matiasgarciaisaia** [comment]
I'm on the go, but do please open a new issue for this, with all the info
you feel relevant. We definitely want to keep improving the release
process, so let's track this properly 👌

**Pull request closed without merging**